crates/sui-types/src/nitro_attestation.rs

```diff
@@ -391,138 +400,215 @@ impl AttestationDocument {
-                for i in [0, 1, 2, 3, 4, 8] {
-                    if let Some((_, ciborium::value::Value::Bytes(val))) = pcr_map.iter().find(
-                        |(k, _)| matches!(k, ciborium::value::Value::Integer(n) if *n == i.into()),
-                    ) {
-                        pcr_vec.push(val.clone());
-                    } else {
+        if let Some(data) = &nonce {
+            if data.len() > MAX_USER_DATA_LENGTH {
+                return Err(NitroAttestationVerifyError::InvalidAttestationDoc(
+                    "invalid nonce".to_string(),
+                ));
+            }
+        }
+
+        let pcrs = document_map
+            .get("pcrs")
+            .ok_or(NitroAttestationVerifyError::InvalidAttestationDoc(
+                "pcrs not found".to_string(),
+            ))?
+            .as_map()
+            .ok_or(NitroAttestationVerifyError::InvalidAttestationDoc(
+                "invalid pcrs format".to_string(),
+            ))
+            .and_then(|pairs| {
+                if pairs.len() > MAX_PCRS_LENGTH {
+                    return Err(NitroAttestationVerifyError::InvalidAttestationDoc(
+                        "invalid PCRs length".to_string(),
+                    ));
+                }
+                let mut pcr_vec = Vec::with_capacity(pairs.len());
+                for (k, v) in pairs.iter() {
+                    let key = k.as_integer().ok_or(
+                        NitroAttestationVerifyError::InvalidAttestationDoc(
+                            "invalid PCR key format".to_string(),
+                        ),
+                    )?;
+                    let value =
+                        v.as_bytes()
+                            .ok_or(NitroAttestationVerifyError::InvalidAttestationDoc(
+                                "invalid PCR value format".to_string(),
+                            ))?;
+
+                    if value.len() != 32 && value.len() != 48 && value.len() != 64 {
                         return Err(NitroAttestationVerifyError::InvalidAttestationDoc(
-                            "invalid PCR format".to_string(),
+                            "invalid PCR value length".to_string(),
                         ));
                     }
-                }
-                pcr_vec
-            }
-            _ => {
-                return Err(NitroAttestationVerifyError::InvalidAttestationDoc(
-                    "cannot parse PCRs".to_string(),
-                ))
-            }
-        };
 
-        let user_data = get_optional_bytes("user_data");
-        let nonce = get_optional_bytes("nonce");
+                    // Valid PCR indices are 0, 1, 2, 3, 4, 8 for AWS.
+                    let key_u64 = u64::try_from(key).map_err(|_| {
+                        NitroAttestationVerifyError::InvalidAttestationDoc(
+                            "invalid PCR index".to_string(),
+                        )
+                    })?;
+
+                    for i in [0, 1, 2, 3, 4, 8] {
```

Comment:
should we return error if we see, e.g., pcr 5?

Reply:
these are actually all valid pcrs, at other places they are just populated as all zeros. we only parse the ones we/aws actually cares about

Reply:
let's add a comment with a link to the docs that say that aws only uses those pcrs